contrib/internal/httptrace/inferred_proxy.go

```diff
@@ -0,0 +1,152 @@
+	ProxyHeaderHttpMethod = "X-Dd-Proxy-Httpmethod"
+
+	// ProxyHeaderDomain is the header used to indicate the AWS domain name
+	// handling the request. This value corresponds to 'context.domainName'
+	// in AWS API Gateway, which represents the custom domain associated
+	// with the API Gateway.
+	ProxyHeaderDomain = "X-Dd-Proxy-Domain-Name"
+
+	// ProxyHeaderStage is the header used to indicate the AWS stage name
+	// for the API request. This value corresponds to 'context.stage' in
+	// AWS API Gateway, and provides the stage (e.g., dev, prod, etc.)
+	// in which the request is being processed.
+	ProxyHeaderStage = "X-Dd-Proxy-Stage"
+)
+
+type proxyDetails struct {
+	spanName  string
+	component string
+}
+
+type proxyContext struct {
+	startTime       time.Time
+	method          string
+	path            string
+	stage           string
+	domainName      string
+	proxySystemName string
+}
+
+var (
+	supportedProxies = map[string]proxyDetails{
+		"aws-apigateway": {
+			spanName:  "aws.apigateway",
+			component: "aws-apigateway",
+		},
+	}
+)
+
+func extractInferredProxyContext(headers http.Header) (*proxyContext, error) {
+	_, exists := headers[ProxyHeaderStartTimeMs]
+	if !exists {
+		return nil, errors.New("proxy header start time does not exist")
+	}
+
+	proxyHeaderSystem, exists := headers[ProxyHeaderSystem]
+	if !exists {
+		return nil, errors.New("proxy header system does not exist")
+	}
+
+	if _, ok := supportedProxies[proxyHeaderSystem[0]]; !ok {
+		return nil, errors.New("unsupported Proxy header system")
+	}
+
+	pc := proxyContext{
+		method:          headers.Get(ProxyHeaderHttpMethod),
+		path:            headers.Get(ProxyHeaderPath),
+		stage:           headers.Get(ProxyHeaderStage),
+		domainName:      headers.Get(ProxyHeaderDomain),
+		proxySystemName: headers.Get(ProxyHeaderSystem),
+	}
+
+	startTimeUnixMilli, err := strconv.ParseInt(headers[ProxyHeaderStartTimeMs][0], 10, 64)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing time string: %v", err)
+	}
+	pc.startTime = time.UnixMilli(startTimeUnixMilli)
+
+	return &pc, nil
+}
+
+func startInferredProxySpan(requestProxyContext *proxyContext, parent ddtrace.SpanContext, opts ...ddtrace.StartSpanOption) tracer.Span {
+	proxySpanInfo := supportedProxies[requestProxyContext.proxySystemName]
+	log.Debug(`Successfully extracted inferred span info ${proxyContext} for proxy: ${proxyContext.proxySystemName}`)
+
+	startTime := requestProxyContext.startTime
+
+	configService := requestProxyContext.domainName
+	if configService == "" {
+		configService = globalconfig.ServiceName()
+	}
+
+	optsLocal := make([]ddtrace.StartSpanOption, len(opts), len(opts)+1)
+	copy(optsLocal, opts)
+
+	optsLocal = append(optsLocal,
+		func(cfg *ddtrace.StartSpanConfig) {
+			if cfg.Tags == nil {
+				cfg.Tags = make(map[string]interface{})
+			}
+
+			cfg.Parent = parent
+			cfg.StartTime = startTime
+
+			cfg.Tags[ext.SpanType] = ext.SpanTypeWeb
+			cfg.Tags[ext.ServiceName] = configService
```

Comment:
What happens if the service name is not provided? 

The JS version tries to use the domain name passed on before using it as the service name: https://github.com/DataDog/dd-trace-js/pull/4837/files .

Reply:
We do the same thing: https://github.com/DataDog/dd-trace-go/blob/d64e3f81223b4a41091120c9a3edc02eb6e511ae/contrib/internal/httptrace/inferred_proxy.go#L121

Use the service name if not passed in